docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,568 @@
+
+The control plane covered by this document is defined as the Kubernetes API server, scheduler, controller manager, DNS
+and proxy services, and the underlying etcd data store.
+
+## Motivation
+
+During 2019 we saw control plane management implementations in each infrastructure provider. Much like
+bootstrapping was identified as being reimplemented in every infrastructure provider and then extracted into Cluster API
+Bootstrap Provider Kubeadm (CABPK), we believe we can reduce the redundancy of control plane management across providers
+and centralize the logic in Cluster API.
+
+### Goals
+
+- To establish new resource types for control plane management
+- To support single node and multiple control plane instances, with the requirement that the infrastructure provider supports some type of a stable endpoint for the API Server (Load Balancer, VIP, etc).
+- To enable declarative orchestrated control plane upgrades
+- To provide a default machine-based implementation using kubeadm
+
+#### Additional goals of the default kubeadm machine-based Implementation
+
+- To provide a kubeadm-based implementation that is infrastructure provider agnostic
+- To enable declarative orchestrated replacement of control plane machines, such as to rollout an OS-level CVE fix.
+- To manage a kubeadm-based, "stacked etcd" control plane
+- To enable scaling of the number of control plane nodes
+- To support pre-existing, user-managed, external etcd clusters
+- To support only user-initiated remediation:
+  E.g. user deletes a Machine. Control Plane Provider reconciles by removing the corresponding etcd member and updating related metadata
+
+### Non-Goals / Future Work
+
+Non-Goals listed in this document are intended to scope bound the current v1alpha3 implementation and are subject to change based on user feedback over time.  
+
+- To manage non-machine based topologies, e.g.
+  - Pod based control planes; these can be managed via standard kubernetes objects.
+  - Non-node control planes (i.e. EKS, GKE, AKS); these can be managed via the respective APIs.
+- To manage control plane deployments across failure domains, followup work for this will be tracked on [this issue](https://github.com/kubernetes-sigs/cluster-api/issues/1647).
+- To manage CA certificates outside of what is provided by Kubeadm bootstrapping
+- To manage etcd clusters in any topology other than stacked etcd.
+- To address disaster recovery constraints, e.g. restoring a control plane from 0 replicas using a filesystem or volume snapshot copy of data persisted in etcd.
+- To support rollbacks, as there is no datastore rollback guarantee for Kubernetes. Consumers should perform backups of the cluster prior to performing potentially destructive operations.
+- To mutate the configuration of live, running clusters (e.g. changing api-server flags), as this is the responsibility of the [component configuration working group](https://github.com/kubernetes/community/tree/master/wg-component-standard).
```

Comment:
Will we have something in place to block this? We ran into a scenario where we needed to do this in v1alpha2. 

If we don't, it seems like it would be possible to change config flags by updating the `KubeadmConfigSpec` and (manually) rolling the control plane nodes one by one.

Reply:
@alexbrand currently we are adding a requirement that the `KubeadmConfigSpec` be treated immutably: https://github.com/kubernetes-sigs/cluster-api/pull/1613/files#diff-32997217ac55782fca9596683679c2bfR308

We've also called this out as part of the Risks: https://github.com/kubernetes-sigs/cluster-api/pull/1613/files#diff-32997217ac55782fca9596683679c2bfR542

Reply:
Gotcha! Thanks, @detiber. It seems like we'd need to create a new cluster if the KubeadmConfig required changes. I have to catch up on the ComponentConfig efforts as well.

Reply:
My main motivation here is to keep the initial implementation a bit simplified. We'll need to figure out how to detect "safe" vs "unsafe" changes in the future or a way to detect safe rollback for "configuration" or "base image" only changes where kubernetes version is not affected.

Reply:
Yes, absolutely. That makes sense to me 👍 .